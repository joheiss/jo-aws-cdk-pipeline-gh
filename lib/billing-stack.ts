import { Stack, StackProps } from 'aws-cdk-lib';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';
import { CostBudgetWarning } from './constructs/budget-construct';

export class BillingStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // create SNS topic for budget warning
    const topic =  new Topic(this, 'MonthlyCostBudgetWarningTopic', {
      displayName: 'Monthly Cost Budget Warning Topic',
      topicName: 'MonthlyCostBudgetWarningTopic',
  });

    // create budget warning 
    const budgetWarning = new CostBudgetWarning(this, 'Budget-Warning', {
      name: 'MonthlyCostBudgetWarning',
      limitAmount: 10,
      snsTopicArn: topic.topicArn,
    });

  }
}
