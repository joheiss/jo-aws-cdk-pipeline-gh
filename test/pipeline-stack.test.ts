import { App, Stack } from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";
import { IStage } from "aws-cdk-lib/aws-codepipeline";
import { PipelineStack } from "../lib/pipeline-stack";
import { ServiceStack } from "../lib/service-stack";
import { BillingStack } from "../lib/billing-stack";

const testEnv = {
  account: '123456789',
  region: 'us-east-1'
};

describe("Pipeline Stack", () => {
  let app: App;
  let stack: PipelineStack;

  beforeEach(() => {
    app = new App();
    stack = new PipelineStack(app, "PipelineStack", {
      env: testEnv,
      ownerEmail: "test@test.de",
    });
  });

  test("Stack", () => {
    const cfn = Template.fromStack(stack).toJSON();
    expect(cfn).toMatchSnapshot();
  });

  describe("Deployment Stage", () => {
    let serviceStack: ServiceStack;
    let stage: IStage;

    beforeEach(() => {
      serviceStack = new ServiceStack(app, "ServiceStack", {
        env: testEnv,
        stageName: "Test",
      });
      stage = stack.addDeployStage(serviceStack, "Deploy");
    });

    test("Add service stage", () => {
      // GIVEN & WHEN -- see beforeEach
      // THEN
      Template.fromStack(stack).hasResourceProperties(
        "AWS::CodePipeline::Pipeline",
        {
          Stages: Match.arrayWith([Match.objectLike({ Name: "Deploy" })]),
        }
      );
    });

    test("Add billing stack to deploy stage", () => {
      // GIVEN & WHEN -- see beforeEach
      // WHEN
      const billingStack = new BillingStack(app, "BillingStack");
      stack.addBillingStackToDeployStage(billingStack, stage);
      // THEN
      Template.fromStack(stack).hasResourceProperties(
        "AWS::CodePipeline::Pipeline",
        {
          Stages: Match.arrayWith([
            Match.objectLike({
              Actions: Match.arrayWith([
                Match.objectLike({ Name: "BudgetWarning_Update" }),
              ]),
            }),
          ]),
        }
      );
    });
  });
});
