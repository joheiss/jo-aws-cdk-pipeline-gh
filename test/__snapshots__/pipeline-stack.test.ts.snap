// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline Stack Stack 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "JoCdkBuildProjectE31CC639": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "CODEPIPELINE",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "JoCdkBuildProjectRole69A5B0CB",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "buildspecs/cdk-buildspec.yml",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "JoCdkBuildProjectRole69A5B0CB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "JoCdkBuildProjectRoleDefaultPolicyCFF13302": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "JoCdkBuildProjectE31CC639",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "JoCdkBuildProjectE31CC639",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:123456789:report-group/",
                    Object {
                      "Ref": "JoCdkBuildProjectE31CC639",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineArtifactsBucket2C0F1ECF",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "JoPipelineArtifactsBucket2C0F1ECF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JoCdkBuildProjectRoleDefaultPolicyCFF13302",
        "Roles": Array [
          Object {
            "Ref": "JoCdkBuildProjectRole69A5B0CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JoPipeline0A0A993E": Object {
      "DependsOn": Array [
        "JoPipelineRoleDefaultPolicyCB6C2CEE",
        "JoPipelineRole33738F64",
      ],
      "Properties": Object {
        "ArtifactStore": Object {
          "Location": Object {
            "Ref": "JoPipelineArtifactsBucket2C0F1ECF",
          },
          "Type": "S3",
        },
        "Name": "JoPipeline",
        "RestartExecutionOnUpdate": true,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "JoPipelineRole33738F64",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "ThirdParty",
                  "Provider": "GitHub",
                  "Version": "1",
                },
                "Configuration": Object {
                  "Branch": "main",
                  "OAuthToken": "{{resolve:secretsmanager:github-token:SecretString:::}}",
                  "Owner": "joheiss",
                  "PollForSourceChanges": false,
                  "Repo": "jo-aws-cdk-pipeline-gh",
                },
                "Name": "Pipeline_Source",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "CdkSourceOutput",
                  },
                ],
                "RunOrder": 1,
              },
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "ThirdParty",
                  "Provider": "GitHub",
                  "Version": "1",
                },
                "Configuration": Object {
                  "Branch": "main",
                  "OAuthToken": "{{resolve:secretsmanager:github-token:SecretString:::}}",
                  "Owner": "joheiss",
                  "PollForSourceChanges": false,
                  "Repo": "jo-aws-express-app",
                },
                "Name": "Service_Source",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "ServiceSourceOutput",
                  },
                ],
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "JoCdkBuildProjectE31CC639",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "CdkSourceOutput",
                  },
                ],
                "Name": "Pipeline_Build",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "CdkBuildOutput",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineBuildPipelineBuildCodePipelineActionRole25B1FBB1",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "JoServiceBuildProjectAA647C41",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "ServiceSourceOutput",
                  },
                ],
                "Name": "Service_Build",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "ServiceBuildOutput",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineBuildServiceBuildCodePipelineActionRoleA9B8ABD3",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CloudFormation",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ActionMode": "CREATE_UPDATE",
                  "Capabilities": "CAPABILITY_NAMED_IAM",
                  "RoleArn": Object {
                    "Fn::GetAtt": Array [
                      "JoPipelineUpdatePipelineUpdateRoleE01C7BDB",
                      "Arn",
                    ],
                  },
                  "StackName": "PipelineStack",
                  "TemplatePath": "CdkBuildOutput::PipelineStack.template.json",
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "CdkBuildOutput",
                  },
                ],
                "Name": "Pipeline_Update",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineUpdatePipelineUpdateCodePipelineActionRole49C2F43C",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Update",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "JoPipelineArtifactsBucket2C0F1ECF": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "JoPipelineArtifactsBucketPolicyEF956EA5": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "JoPipelineArtifactsBucket2C0F1ECF",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineArtifactsBucket2C0F1ECF",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "JoPipelineArtifactsBucket2C0F1ECF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "JoPipelineBuildPipelineBuildCodePipelineActionRole25B1FBB1": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "JoPipelineBuildPipelineBuildCodePipelineActionRoleDefaultPolicy34EC7937": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JoCdkBuildProjectE31CC639",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JoPipelineBuildPipelineBuildCodePipelineActionRoleDefaultPolicy34EC7937",
        "Roles": Array [
          Object {
            "Ref": "JoPipelineBuildPipelineBuildCodePipelineActionRole25B1FBB1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JoPipelineBuildServiceBuildCodePipelineActionRoleA9B8ABD3": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "JoPipelineBuildServiceBuildCodePipelineActionRoleDefaultPolicy6A838C5E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JoServiceBuildProjectAA647C41",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JoPipelineBuildServiceBuildCodePipelineActionRoleDefaultPolicy6A838C5E",
        "Roles": Array [
          Object {
            "Ref": "JoPipelineBuildServiceBuildCodePipelineActionRoleA9B8ABD3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JoPipelineRole33738F64": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "JoPipelineRoleDefaultPolicyCB6C2CEE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineArtifactsBucket2C0F1ECF",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "JoPipelineArtifactsBucket2C0F1ECF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JoPipelineBuildPipelineBuildCodePipelineActionRole25B1FBB1",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JoPipelineBuildServiceBuildCodePipelineActionRoleA9B8ABD3",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JoPipelineUpdatePipelineUpdateCodePipelineActionRole49C2F43C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JoPipelineRoleDefaultPolicyCB6C2CEE",
        "Roles": Array [
          Object {
            "Ref": "JoPipelineRole33738F64",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JoPipelineSourcePipelineSourceWebhookResource7F6FB57B": Object {
      "Properties": Object {
        "Authentication": "GITHUB_HMAC",
        "AuthenticationConfiguration": Object {
          "SecretToken": "{{resolve:secretsmanager:github-token:SecretString:::}}",
        },
        "Filters": Array [
          Object {
            "JsonPath": "$.ref",
            "MatchEquals": "refs/heads/{Branch}",
          },
        ],
        "RegisterWithThirdParty": true,
        "TargetAction": "Pipeline_Source",
        "TargetPipeline": Object {
          "Ref": "JoPipeline0A0A993E",
        },
        "TargetPipelineVersion": 1,
      },
      "Type": "AWS::CodePipeline::Webhook",
    },
    "JoPipelineSourceServiceSourceWebhookResource9B621683": Object {
      "Properties": Object {
        "Authentication": "GITHUB_HMAC",
        "AuthenticationConfiguration": Object {
          "SecretToken": "{{resolve:secretsmanager:github-token:SecretString:::}}",
        },
        "Filters": Array [
          Object {
            "JsonPath": "$.ref",
            "MatchEquals": "refs/heads/{Branch}",
          },
        ],
        "RegisterWithThirdParty": true,
        "TargetAction": "Service_Source",
        "TargetPipeline": Object {
          "Ref": "JoPipeline0A0A993E",
        },
        "TargetPipelineVersion": 1,
      },
      "Type": "AWS::CodePipeline::Webhook",
    },
    "JoPipelineUpdatePipelineUpdateCodePipelineActionRole49C2F43C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "JoPipelineUpdatePipelineUpdateCodePipelineActionRoleDefaultPolicyE0551ED1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JoPipelineUpdatePipelineUpdateRoleE01C7BDB",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineArtifactsBucket2C0F1ECF",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "JoPipelineArtifactsBucket2C0F1ECF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "cloudformation:CreateStack",
                "cloudformation:DescribeStack*",
                "cloudformation:GetStackPolicy",
                "cloudformation:GetTemplate*",
                "cloudformation:SetStackPolicy",
                "cloudformation:UpdateStack",
                "cloudformation:ValidateTemplate",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":cloudformation:us-east-1:123456789:stack/PipelineStack/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JoPipelineUpdatePipelineUpdateCodePipelineActionRoleDefaultPolicyE0551ED1",
        "Roles": Array [
          Object {
            "Ref": "JoPipelineUpdatePipelineUpdateCodePipelineActionRole49C2F43C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JoPipelineUpdatePipelineUpdateRoleDefaultPolicyA30865F5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineArtifactsBucket2C0F1ECF",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "JoPipelineArtifactsBucket2C0F1ECF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JoPipelineUpdatePipelineUpdateRoleDefaultPolicyA30865F5",
        "Roles": Array [
          Object {
            "Ref": "JoPipelineUpdatePipelineUpdateRoleE01C7BDB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JoPipelineUpdatePipelineUpdateRoleE01C7BDB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudformation.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "JoServiceBuildProjectAA647C41": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "CODEPIPELINE",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "JoServiceBuildProjectRole7B69B4F4",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "buildspecs/service-buildspec.yml",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "JoServiceBuildProjectRole7B69B4F4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "JoServiceBuildProjectRoleDefaultPolicy543D32ED": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "JoServiceBuildProjectAA647C41",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456789:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "JoServiceBuildProjectAA647C41",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:123456789:report-group/",
                    Object {
                      "Ref": "JoServiceBuildProjectAA647C41",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "JoPipelineArtifactsBucket2C0F1ECF",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "JoPipelineArtifactsBucket2C0F1ECF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JoServiceBuildProjectRoleDefaultPolicy543D32ED",
        "Roles": Array [
          Object {
            "Ref": "JoServiceBuildProjectRole7B69B4F4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "NotifyOnFailedPipeline67CBA9FC": Object {
      "Type": "AWS::SNS::Topic",
    },
    "NotifyOnFailedPipelinetesttestdeFF5239F1": Object {
      "Properties": Object {
        "Endpoint": "test@test.de",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "NotifyOnFailedPipeline67CBA9FC",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
